transport/http2_client.go

```diff
@@ -1258,12 +1286,15 @@ func (t *http2Client) itemHandler(i item) error {
 	case *windowUpdate:
 		err = t.framer.fr.WriteWindowUpdate(i.streamID, i.increment)
 	case *settings:
-		if i.ack {
-			t.applySettings(i.ss)
+		err = t.framer.fr.WriteSettings(i.ss...)
+	case *settingsAck:
+		if i.rs == nil && i.ps == nil {
 			err = t.framer.fr.WriteSettingsAck()
-		} else {
-			err = t.framer.fr.WriteSettings(i.ss...)
+			break
 		}
+		t.applySettings(i.rs)
+		t.controlBuf.put(&settingsAck{})
```

Comment:
Why not apply settings before enqueuing the ack the first time?

This results in more delay in processing the ack (it has to make it through the control buffer twice), and it splits up the code that handles settings, which makes it harder to reason about.

If there's some reason we have to do it this way, that's fine but we should put comments in to explain it.

Reply:
I was in fact thinking about that last night. The code sort of evolved to be this way and you're right it makes more sense to apply the settings in the settings handler itself.